test/parallel/test-tls-psk-circuit.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+const common = require('../common');
+
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const assert = require('assert');
+const tls = require('tls');
+
+const CIPHERS = 'PSK+HIGH';
+const USERS = {
+  UserA: Buffer.allocUnsafe(128),
+  UserB: Buffer.from('82072606b502b0f4025e90eb75fe137d', 'hex'),
+};
+const CLIENT_IDENTITIES = [
+  { psk: USERS.UserA, identity: 'UserA' },
+  { psk: USERS.UserB, identity: 'UserB' },
+  // Unrecognized user should fail handshake
+  { psk: USERS.UserB, identity: 'UserC' },
+  // Recognized user but incorrect secret should fail handshake
+  { psk: USERS.UserA, identity: 'UserB' },
+  { psk: USERS.UserB, identity: 'UserB' }
+];
+const TEST_DATA = 'Hello from test!';
+
+const clientResults = [];
+const connectingIds = [];
+
+const serverOptions = {
+  ciphers: CIPHERS,
+  pskCallback(id) {
+    connectingIds.push(id);
+    if (USERS.hasOwnProperty(id)) {
+      return USERS[id];
+    }
+  },
+};
+
+const server = tls.createServer(serverOptions, common.mustCall((c) => {
+  c.once('data', (data) => {
+    assert.strictEqual(data.toString(), TEST_DATA);
+    c.end(data);
+  });
+}, 3));
+
+server.listen(0, () => {
+  startTest(server.address().port);
+});
+
+function startTest(port) {
+  function connectClient(options, next) {
+    const client = tls.connect(port, 'localhost', options, () => {
+      clientResults.push('connect');
+
+      client.end(TEST_DATA);
+
+      client.on('data', common.mustCallAtLeast((data) => {
+        assert.strictEqual(data.toString(), TEST_DATA);
+      }));
+
+      client.on('close', common.mustCall(next));
+    });
+
+    let errored = false;
+    client.on('error', (err) => {
+      if (!errored) {
+        clientResults.push(err.message);
+        errored = true;
+        next();
+      }
+    });
+  }
+
+  function doTestCase(tcnum) {
+    if (tcnum >= CLIENT_IDENTITIES.length) {
+      server.close();
+    } else {
+      connectClient({
+        ciphers: CIPHERS,
+        checkServerIdentity: () => {},
+        pskCallback: () => CLIENT_IDENTITIES[tcnum]
+      }, () => {
+        doTestCase(tcnum + 1);
+      });
+    }
+  }
+  doTestCase(0);
+}
+
+const DISCONNECT_MESSAGE =
+  'Client network socket disconnected before ' +
+  'secure TLS connection was established';
+
+process.on('exit', () => {
```

Comment:
We have lots of testcases that use this pattern, but they are our of style.  It would be better to fire off the tests with common.mustCall(), thus guaranteeing they will occur and not silently fail and have node exit with `0`. Each test would know its expected result (sucess/fail), and assert the expectations when complete, rather than pushing the results into global data and checking the expectations on exit. This also allows the tests to be run in parallel rather than sequentially, since each is self-contained. #25409 is an example of transforming from sequential check-on-exit style to parallel mustCall() style.

Reply:
That's a great idea. I kind of stuck with what was here initially but having properly separate test cases is a lot better.